source/common/ssl/context_impl.cc

```diff
@@ -137,35 +147,53 @@ bssl::UniquePtr<SSL> ContextImpl::newSsl() const {
   return bssl::UniquePtr<SSL>(SSL_new(ctx_.get()));
 }
 
-bool ContextImpl::verifyPeer(SSL* ssl) const {
-  bool verified = true;
+int ContextImpl::verifyCallback(int ok, X509_STORE_CTX* store) {
```

Comment:
So, to be honest, I don't understand the legacy X.509 code. We don't maintain it (just sync from upstream) and are ultimately moving away from it:
https://boringssl.googlesource.com/boringssl/+/master/include/openssl/x509.h#92

This callback is threaded pretty deep into that verify function and is partially intended to let you override errors and the like. It gets called in lots of places.
https://www.openssl.org/docs/man1.0.2/crypto/X509_STORE_CTX_set_verify_cb.html

Checking the error_depth, in particular, is puzzling to me, though it probably works? I would expect you to merely do:
```
// Don't override any errors.
if (!ok) {
  return ok;
}

// Go do your thing.
return verifyCertificate, etc.
```
Another thing you can do, which might be more straightforward is use `SSL_CTX_set_cert_verify_callback`. It just swaps out the `X509_verify_cert` call altogether.
https://commondatastorage.googleapis.com/chromium-boringssl-docs/ssl.h.html#SSL_CTX_set_cert_verify_callback
Then you could do something like:

```
int ret = X509_verify_cert(ctx);
if (ret <= 0) {
  return ret;
}

// Go do your thing.
return verifyCertificate(....);
```